### PR TITLE
Add ConnectorMetadata#isMaterializedView

### PR DESCRIPTION
**Pull request by ksobolew**
## Description

The current implementation of `MetadataManager#isMaterializedView` calls `ConnectorMetadata#getMaterializedView` under the covers. This means that we have to load and deserialize the materialized view metadata just to see if it exists, which is unnecessary most of the time. This may also cause issues when the metadata is currupted in some way (see the last commit for relevant test cases, where we add support for this method to the Iceberg plugin). In such a case, we want to still be able to `DROP` the corrupted materialized view - which is currently impossible, because the `DropMaterializedViewTask` will call `getMaterializedView`, which will fail when it tries to load the corrupted metadata. With this change this is less likely.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix failure in `DROP MATERIALIZED VIEW` in Iceberg when the materialized view metadata is missing or corrupted.
```

### CODE REVIEW

**Comment by ksobolew** [comment]
CI #16437

**Comment by findepi** [comment]
> And that would prevent it from being `DROP`ped, because the `DropMaterializedViewTask` first wants to check if the entity to be dropped is a materialized view. If it called `getMaterializedView`, as it used to, dropping such a view would fail too; but with `isMaterializedView` that is not a problem.

please add a test for that

> This may not be just an optimization,

is this **also** an optimization? if so, there should be a test change (eg counting file system or metastore access) that shows the improvement.

**Comment by findepi** [comment]
> (x) This is not user-visible or is docs only, and no release notes are required.

The description contradicts that.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 1]
```diff
@@ -1552,6 +1552,21 @@ default Optional<ConnectorMaterializedViewDefinition> getMaterializedView(Connec
         return Optional.empty();
     }
 
+    /**
+     * Tells whether the specified view name refers to a materialized view. Returns {@literal false}
+     * if {@code viewName} relation does not exist or is not a materialized view (e.g. is a table, or a view).
+     * The default implementation calls {@link #getMaterializedView(ConnectorSession, SchemaTableName)} and
+     * returns {@literal true} if and only if that method returns a non-empty {@link Optional}. The implementing
+     * class may provide a more optimized implementation, which does not need to create the
+     * {@link ConnectorMaterializedViewDefinition} instance unnecessarily.
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1561]: Remove

**Comment by ksobolew** [comment]
> is this **also** an optimization? if so, there should be a test change (eg counting file system or metastore access) that shows the improvement.

There are already changes to the `TestTrinoDatabaseMetaData` that show changes in the call patterns. It's not exactly fewer calls, but fewer to `getMaterializedView`.

**Comment by findepi** [comment]
TestTrinoDatabaseMetaData is quite tied to JDBC driver
can we have a test for Iceberg connector in the Iceberg module as well? 
I think it would make sense, given that Iceberg is the only connector that has MV support

see eg https://github.com/trinodb/trino/blob/8f25ce5e42f5a1a22c238f1b85dd093ee1a5f3c7/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java#L328-L372

**Comment by ksobolew** [comment]
Sure, I'm working on tests right now (though demonstrating how a corrupted metadata can prevent `DROP MV` from executing is kinda challenging)

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java (hunk 1 above)
Comment (ksobolew) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java]: Done

**Comment by ksobolew** [comment]
@findepi Addressed comments, PTAL.

**Comment by ksobolew** [comment]
> TestTrinoDatabaseMetaData is quite tied to JDBC driver
> can we have a test for Iceberg connector in the Iceberg module as well?
> I think it would make sense, given that Iceberg is the only connector that has MV support

Added such a test, but it's not very meaningful, IMO. It just says that we call `getTable` on the metastore one time less when we `DROP MATERIALIZED VIEW`.

**Comment by ksobolew** [comment]
> > (x) This is not user-visible or is docs only, and no release notes are required.
> 
> The description contradicts that.

Added a release note.

**Comment by ksobolew** [comment]
> > And that would prevent it from being `DROP`ped, because the `DropMaterializedViewTask` first wants to check if the entity to be dropped is a materialized view. If it called `getMaterializedView`, as it used to, dropping such a view would fail too; but with `isMaterializedView` that is not a problem.
> 
> please add a test for that

Finally found a way to do that; test added.

> > This may not be just an optimization,
> 
> is this **also** an optimization? if so, there should be a test change (eg counting file system or metastore access) that shows the improvement.

I rewrote the commit message (and the PR description) to phrase it as a primarily correctness issue, with some added efficiency improvements added as a bonus.

**Comment by findepi** [comment]
/test-with-secrets sha=f5d848b3678109c4fe1e67b18b798d5c0e1ee04e

**Comment by findepi** [comment]
i plan to review this (later)

**Comment by findepi** [comment]
> This means that we have to load and deserialize the materialized view metadata just to see if it exists, which is unnecessary most of the time

This is true, but actually it is probably much lesser problem than it sounds it is


>. This may also cause issues when the metadata is currupted in some way (see the last commit for relevant test cases, where we add support for this method to the Iceberg plugin). In such a case, we want to still be able to `DROP` the corrupted materialized view

The `ConnectorMaterializedViewDefinition` is the "materialized view definition", it shouldn't need to load the storage table for anything.
I think the fact that in Iceberg case it loads properties from the storage table is sub-optimal design and we should revisit it.

- Alex's https://github.com/trinodb/trino/pull/18853's will squash MV and storage table within metastore
- plus, we should preserve user-visible MV properties in the metastore level object as well
   - we can do so safely, since storage table is "owned" by the MV so it must not change behind the scenes in any way

This is important not only for `DROP corrupted_mv` case.
It is also important for all metadata queries. Querying `system.metatda.materialized_views` should ideally not need to go to S3 to obtain "materialized view definitions", especially not O(number of MVs) times.

**Comment by ksobolew** [comment]
> This is true, but actually it is probably much lesser problem than it sounds it is

It is a problem, at least, for the use case I'm trying to fix.

> This is important not only for DROP corrupted_mv case.
> It is also important for all metadata queries. Querying system.metatda.materialized_views should ideally not need to go to S3 to obtain "materialized view definitions", especially not O(number of MVs) times.

That's true as well.

I appreciate the feedback and I agree that we don't want to add more code which may be considered a workaround for limitations of the current architecture, but I also don't want this to become a case of "sure, but first please spend 6 months refactoring this complex piece of code". What would be the commended course of action then?

**Comment by mosabua** [comment]
Reminder to continue work on this @ksobolew @findepi or close it.

**Comment by ksobolew** [comment]
This work is currently on hold and it's not clear this is the way forward. I'll reopen if it crystalizes.

**Pull request closed without merging**